Support MetricNameSpace in executors and include shuffle to shuffleservice metrics

matyix opened the pull request:
## What changes were proposed in this pull request?
Support MetricNameSpace in `shuffle` to shuffleservice metrics.

Pass `spark.metrics.namespace` configuration to shuffle such as the shuffleservice can publish metrics with custom namespace. In case of shuffleservice the `shuffle` prefix is used in the metrics.

## How was this patch tested?

Executed all unit and integration tests.

Manual testing through deploying a Spark cluster, Prometheus server, Pushgateway and ran SparkPi - and checking for the shuffle metrics in Pushgateway/Grafana.

matyix committed:
Support MetricNameSpace in executors and include shuffle to shuffleservice metrics

liyinan926 commented:
Unit test build for this PR seems stuck for almost 3 days.

foxish commented:
@ssuchter @kimoonkim, should the unit test build timeout eventually?

liyinan926 commented:
Some test failed due to OOM. Can someone who has admin access to the Jenkins instance kill the build?

kimoonkim commented:
@foxish I think we can time out the Jenkins jobs, say, after 8 hours. I'll see if i can make this change to Jenkins jobs.

@liyinan926 Yes, let me find and kill the hanging build.

kimoonkim commented:
I just checked. The build was killed already by Jenkins. The build log [console](http://spark-k8s-jenkins.pepperdata.org:8080/job/PR-spark-k8s-unit-tests/1111/console]) says:
```
...
Build was aborted
Aborted by Jenkins Admin
```

kimoonkim commented:
Hmm, also the unit test had set up timeout already. The config page says:
```
Abort the build if it's stuck
Time-out strategy Absolute
    Timeout minutes: 60
```

I guess it doesn't always work :-(

cvpatel commented:
@liyinan926 @kimoonkim @foxish Sorry forgot to update, but I killed the test earlier and added a 60 minute timeout to the configuration around noon today.

cvpatel commented:
rerun all tests please

matyix committed:
Merge branch 'branch-2.2-kubernetes' into metric-namespace-support

matyix committed:
Merge branch 'branch-2.2-kubernetes' into metric-namespace-support

mccheah commented:
I'm a little confused as to what this property actually does. I see the configuration of `spark.metrics.namespace` being set, but I don't see it being used in this pull request itself?

mccheah commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -201,7 +202,8 @@ private[spark] object CoarseGrainedExecutorBackend extends Logging {
       val cfg = driver.askSync[SparkAppConfig](RetrieveSparkAppConfig(executorId))
-      val props = cfg.sparkProperties ++ Seq[(String, String)](("spark.app.id", appId))
+      val props = cfg.sparkProperties ++ Seq[(String, String)](("spark.app.id", appId),
+        ("spark.metrics.namespace", metricsNamespace))
```
Why can't the driver provide this as it would with any other Spark property? i.e. if I set `spark.metrics.namespace` in my SparkConf then shouldn't that be what's used here?

matyix committed:
Revert passing metrics namespace to executores via environment variable as that is pulled from driver spark conf

matyix commented:
Hello @mccheah - thanks for the feedback and review. The starting base for the PR was an earlier version and the custom metrics namespace in executors was missing back then. I have reverted the changes and updated the PR text as well reflecting the latest commit.

matyix committed:
Remove unused metrics namespace from Dockerfiles

erikerlandson commented:
Is the idea here to backport the spark support for custom metric namespaces to this fork?  I ask because I'd expect this to be resolved either via the upstreaming process or via our next rebase.

matyix commented:
@erikerlandson After rebasing with latest branch-2.2-kubernetes the original PR has reduced the scope to support custom metrics namespace for `external shuffle service` only (missing from master).

erikerlandson commented:
@matyix Oh I see - I was referring to the next time we rebase against an upstream release (spark-2.2.1 or spark-2.3, etc)

matyix commented:
@erikerlandson shall I push this PR upstream as well?

matyix commented:
Closing this as it's redundant, this PR https://github.com/apache/spark/pull/19775 fixes this one as well.
